docs/source/building/variants.rst

```diff
@@ -0,0 +1,1230 @@
+an ``extend_keys`` entry.  If you list it in only one of them, an error will be
+raised, to avoid confusion with one conda_build_config.yaml file that would add
+entries to the build matrix, and another which would not. For example, this
+should raise an error:
+
+.. code-block:: yaml
+
+   some_trait:
+     - dog
+
+
+recipe/conda_build_config.yaml:
+
+.. code-block:: yaml
+
+   some_trait:
+     - pony
+   extend_keys:
+     - some_trait
+
+
+When our two proper yaml config files are combined, ordinarily the recipe-local
+variant would clobber the user-wide variant, yielding ``{'some_trait':
+'pony'}``. However, with the extend_keys entry, we end up with what we've always
+wanted: a dog *and* pony show: ``{'some_trait': ['dog', 'pony'])}``
+
+Again, mostly an internal implementation detail - unless you find a use for it.
+Internally, it is used to aggregate the ``pin_run_as_build`` and
+``ignore_version`` entries from any of your conda_build_config.yaml
+files.
+
+
+Customizing compatibility
+-------------------------
+
+.. _pinning_expressions:
+
+Pinning expressions
+~~~~~~~~~~~~~~~~~~~
+
+
+Pinning expressions are the syntax used to specify how many parts of the version
+to pin. They are by convention strings containing ``x`` characters separated by
+``.``. The number of version parts to pin is simply the number of things that
+are separated by ``.``. For example, ``"x.x"`` pins major and minor version.
+``"x"`` pins only major version.
+
+Wherever pinning expressions are accepted, you can customize both lower and
+upper bounds.
+
+
+.. code-block:: python
+
+    # produces pins like >=1.11.2,<1.12
+    variants = [{'numpy': '1.11', 'pin_run_as_build': {'numpy': {'max_pin': 'x.x'}}}]
+
+
+.. code-block:: python
+
+    # produces pins like >=1.11,<2
+    variants = [{'numpy': '1.11', 'pin_run_as_build': {'numpy': {'min_pin': 'x.x', 'max_pin': 'x'}}}]
+
+
+Pinning at the variant level
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Some packages, such as boost, *always* need to be pinned at runtime to the
+version that was present at build time. For these cases where the need for
+pinning is consistent, pinning at the variant level is a good option.
+Conda-build will automatically pin run requirements to the versions present in
+the build environment when the follow conditions are met:
+
+1. The dependency is listed in the requirements/build section. It can be pinned,
+   but does not need to be.
+2. The dependency is listed by name (no pinning) in the requirements/run section
+3. The ``pin_run_as_build`` key in the variant has a value that is a dictionary,
+   containing a key that matches the dependency name listed in the run
+   requirements. The value should be a dictionary with up to 4 keys:
+   ``min_pin``, ``max_pin``, ``lower_bound``, ``upper_bound``. The first two are
+   pinning expressions. The latter two are version numbers, overriding detection
+   of current version. (defaulting to None).
```

Comment:
@msarahan what defaults to None? All four keys, the latter two, or the dictionary that contains the four keys?